File: android_app/src/main/java/me/timeto/app/ui/color_picker/ColorPickerCustomSheet.kt
package me.timeto.app.ui.color_picker

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Slider
import androidx.compose.material.SliderDefaults
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableFloatState
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.ui.H_PADDING
import me.timeto.app.ui.H_PADDING_HALF
import me.timeto.app.ui.VStack
import me.timeto.app.ui.ZStack
import me.timeto.app.ui.c
import me.timeto.app.ui.squircleShape
import me.timeto.app.toColor
import me.timeto.app.ui.Screen
import me.timeto.app.ui.navigation.LocalNavigationLayer
import me.timeto.shared.ColorRgba
import me.timeto.shared.vm.color_picker.ColorPickerVm

// todo prevent navigation back

@Composable
fun ColorPickerCustomSheet(
    initColorRgba: ColorRgba,
    onDone: (ColorRgba) -> Unit,
) {

    val navigationLayer = LocalNavigationLayer.current

    val (initH, initS, initL) = initColorRgba.toHsl()
    val hue = remember { mutableFloatStateOf(initH) }
    val saturation = remember { mutableFloatStateOf(initS) }
    val lightness = remember { mutableFloatStateOf(initL) }

    val colorRgba: ColorRgba =
        ColorRgba.fromHsl(hue.floatValue, saturation.floatValue, lightness.floatValue, initColorRgba.a)

    VStack {

        ZStack(Modifier.weight(1f))

        // todo divider or sheet background color

        Screen(
            modifier = Modifier
                .height(300.dp),
        ) {

            ZStack(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 16.dp)
                    .padding(horizontal = H_PADDING_HALF),
            ) {

                Text(
                    text = "Cancel",
                    modifier = Modifier
                        .align(Alignment.CenterStart)
                        .clip(squircleShape)
                        .clickable {
                            navigationLayer.close()
                        }
                        .padding(
                            horizontal = H_PADDING_HALF,
                            vertical = 4.dp,
                        ),
                    color = c.blue,
                )

                Text(
                    text = ColorPickerVm.prepCustomColorRgbaText(
                        colorRgba = colorRgba,
                    ),
                    modifier = Modifier
                        .align(Alignment.Center)
                        .clip(squircleShape)
                        .background(colorRgba.toColor())
                        .padding(
                            horizontal = 8.dp,
                            vertical = 4.dp,
                        ),
                    color = c.white,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Light,
                )

                Text(
                    text = "Done",
                    modifier = Modifier
                        .align(Alignment.CenterEnd)
                        .clip(squircleShape)
                        .clickable {
                            onDone(colorRgba)
                            navigationLayer.close()
                        }
                        .padding(
                            horizontal = H_PADDING_HALF,
                            vertical = 4.dp,
                        ),
                    color = c.blue,
                    fontWeight = FontWeight.SemiBold,
                )
            }

            HslSliderView(
                value = hue,
                valueRange = 0f..360f,
                color = Color.hsl(hue.floatValue, 1f, 0.5f),
            )

            HslSliderView(
                value = saturation,
                valueRange = 0f..100f,
                color = Color.hsl(hue.floatValue, saturation.floatValue / 100f, 0.5f),
            )

            HslSliderView(
                value = lightness,
                valueRange = 0f..100f,
                color = Color.hsl(hue.floatValue, 1f, lightness.floatValue / 100f),
            )

            ZStack(Modifier.navigationBarsPadding())
        }
    }
}

///

@Composable
private fun HslSliderView(
    value: MutableFloatState,
    valueRange: ClosedFloatingPointRange<Float>,
    color: Color,
) {
    Slider(
        value = value.floatValue,
        onValueChange = { value.floatValue = it },
        modifier = Modifier
            .padding(horizontal = H_PADDING - 6.dp),
        valueRange = valueRange,
        colors = SliderDefaults.colors(
            thumbColor = color,
            activeTrackColor = color,
        ),
    )
}

File: ios_app/iphone/UI/ColorPicker/ColorPickerCustomSheet.swift
import SwiftUI
import shared

struct ColorPickerCustomSheet: View {
    
    init(
        initColorRgba: ColorRgba,
        onDone: @escaping (ColorRgba) -> Void
    ) {
        self.onDone = onDone
        let hsl = initColorRgba.toHsl()
        _colorHsl = State(initialValue: ColorHslLocal(
            h: Double(hsl.h),
            s: Double(hsl.s),
            l: Double(hsl.l)
        ))
    }
    
    ///
    
    @State private var colorHsl: ColorHslLocal
    private let onDone: (ColorRgba) -> Void
    
    @Environment(\.dismiss) private var dismiss
    
    private var colorRgba: ColorRgba {
        ColorRgba.companion.fromHsl(h: Float(colorHsl.h), s: Float(colorHsl.s), l: Float(colorHsl.l), a: 255)
    }
    
    var body: some View {
        VStack(alignment: .center) {
            Spacer()
            ColorSliderView(
                color: Color(hue: colorHsl.h / 360.0, saturation: 1.0, brightness: 1.0),
                value: $colorHsl.h,
                range: 0...360
            )
            ColorSliderView(
                color: Color(hue: colorHsl.h / 360.0, saturation: colorHsl.s / 100.0, brightness: 0.5 + colorHsl.s / 200.0),
                value: $colorHsl.s,
                range: 0...100
            )
            ColorSliderView(
                color: Color(hue: colorHsl.h / 360.0, saturation: 1.0, brightness: colorHsl.l / 100.0),
                value: $colorHsl.l,
                range: 0...100
            )
            Spacer()
        }
        .toolbar {
            ToolbarItem(placement: .cancellationAction) {
                Button("Cancel") {
                    dismiss()
                }
            }
            ToolbarItem(placement: .primaryAction) {
                Button("Done") {
                    onDone(colorRgba)
                    dismiss()
                }
                .fontWeight(.semibold)
            }
            ToolbarItem(placement: .principal) {
                Text(ColorPickerVm.companion.prepCustomColorRgbaText(colorRgba: colorRgba))
                    .font(.system(size: 16))
                    .foregroundColor(.white)
                    .lineLimit(1)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 6)
                    .background(
                        RoundedRectangle(cornerRadius: 8, style: .continuous)
                            .fill(colorRgba.toColor())
                    )
            }
        }
        .onChange(of: colorRgba) { old, new in
            if old != new {
                Haptic.softShot()
            }
        }
        .toolbarTitleDisplayMode(.inline)
        .presentationDetents([.height(220)])
        .interactiveDismissDisabled()
    }
}

///

private struct ColorHslLocal: Equatable {
    var h: Double
    var s: Double
    var l: Double
}

private struct ColorSliderView: View {
    
    let color: Color
    @Binding var value: Double
    var range: ClosedRange<Double>
    
    var body: some View {
        Slider(value: $value, in: range)
            .accentColor(color)
            .padding(.horizontal, H_PADDING)
            .padding(.vertical, 6)
    }
}

